app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/CachedTimelineRemoteMediator.kt

```diff
@@ -0,0 +1,138 @@
+/* Copyright 2021 Tusky Contributors
+ *
+ * This file is a part of Tusky.
+ *
+ * This program is free software; you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+ * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with Tusky; if not,
+ * see <http://www.gnu.org/licenses>. */
+
+package com.keylesspalace.tusky.components.timeline.viewmodel
+
+import androidx.paging.ExperimentalPagingApi
+import androidx.paging.LoadType
+import androidx.paging.PagingState
+import androidx.paging.RemoteMediator
+import androidx.room.withTransaction
+import com.google.gson.Gson
+import com.keylesspalace.tusky.components.timeline.Placeholder
+import com.keylesspalace.tusky.components.timeline.toEntity
+import com.keylesspalace.tusky.db.AccountManager
+import com.keylesspalace.tusky.db.AppDatabase
+import com.keylesspalace.tusky.db.TimelineStatusWithAccount
+import com.keylesspalace.tusky.entity.Status
+import com.keylesspalace.tusky.network.MastodonApi
+import com.keylesspalace.tusky.util.dec
+import kotlinx.coroutines.rx3.await
+import retrofit2.HttpException
+
+@ExperimentalPagingApi
+class CachedTimelineRemoteMediator(
+    accountManager: AccountManager,
+    private val api: MastodonApi,
+    private val db: AppDatabase,
+    private val gson: Gson
+) : RemoteMediator<Int, TimelineStatusWithAccount>() {
+
+    private var initialRefresh = false
+
+    private val timelineDao = db.timelineDao()
+    private val activeAccount = accountManager.activeAccount!!
+
+    override suspend fun load(
+        loadType: LoadType,
+        state: PagingState<Int, TimelineStatusWithAccount>
+    ): MediatorResult {
+
+        try {
+            var dbEmpty = false
+            if (!initialRefresh && loadType == LoadType.REFRESH) {
+                val topId = timelineDao.getTopId(activeAccount.id)
+                topId?.let { cachedTopId ->
+                    val statusResponse = api.homeTimeline(
```

Comment:
we do api.homeTimeline(...).await() + unwrap response multiple times, maybe we should wrap it in another method somewhere here

Reply:
I think I prefer to do this later when replacing Rx with coroutines